Update the instructions of customized agent (#1991)
pytorch/fairseq: examples/simultaneous_translation/docs/evaluation.md

# Introduction to the evaluation interface
The simultaneous translation models from the shared task participants are evaluated under a server-client protocol. 

> The participants are required to plug in their own model API in the protocol, and submit a Docker image.

Reviewer: Docker image or Docker file? I think image is what it built but they don't submit the image, just a way for us to build it and run it.

> The server provides information that client needed and evaluate latency and quality, while client sends to translation. 

Reviewer: ```suggestion
The server provides information needed by the client and evaluates latency and quality, while the client sends the translation. 
```

> The server provides information that client needed and evaluate latency and quality, while client sends to translation. 

Reviewer: What kind of information?

> The fairseq toolkit is a dependency but the evaluation process can be applied in an arbitary framework.

Reviewer: ```suggestion
We use the Fairseq toolkit as an example but the evaluation process can be applied in an arbitary framework.
```

```
The `--score-type` can be either `text` or `speech` to evaluation different tasks.

For text models, the `$src` and `$tgt` are the raw source and target text.

> For speech models, please first follow the Data Preparation [here](baseline.md) excecpt binarazation. After preparation, Only `$tgt` is need, which is the json file in the data directory (for eaxample, dev.json)

Reviewer: ```suggestion
For speech models, please first follow the Data Preparation [here](baseline.md) except for the binarization step. After preparation, Only `$tgt` is need, which is the json file in the data directory (for example, dev.json)
```

> The state that server sent to client has the following format

Reviewer: ```suggestion
The state sent to the client by the server has the following format
```

|Action|Content|
|:---:|:---:|
|Request new word (Text)| ```{key: "GET", value: None}```|
|Request new utterence (Speech) | ```{key: "GET", value: {"segment_size": segment_size}}```|
|Predict word "W"| ```{key: "SEND", value: "W"}```|

The core of the client module is the [agent](../eval/agents/agent.py). 

> One can build a customized agent from the abstract class of agent, shown as follow.

Reviewer: ```suggestion
One can build a customized agent from the abstract class of the agent, shown as follows.
```

    def finish_eval(self, states, new_state):
        # Check if evaluation is finished
        # Return True if finished False not
        ...
    

>     def policy(self, states):

Reviewer: Was this a typo?

Author: Yes

 
```

> Here is an exmaple of a customized agent. 

Reviewer: ```suggestion
Here is an example of a customized agent. 
```

First of all, a name needs to be registered. 

> Next, reload the agent functions according to the translation model. 

Reviewer: What do you mean by reload here?

Author: It means reload the python function, do you prefer another expression?

> The functions that need to reload are listed in `MyAgent` class as follow. 

Reviewer: ```suggestion
The functions that need to be reload are listed in the `MyAgent` class as follow. 
```

> Finally copy the agent file to this [directory](../eval/agents) in the local fairseq repository.

Reviewer: ```suggestion
Finally, copy the agent file to this [directory](../eval/agents) in the local fairseq repository.
```

```python
from example.simultaneous_translation.eval.agents import register_agent
@register_agent("my_agent_name")
class MyAgent(Agent):
    @staticmethod
    def add_args(parser):
        # add customized arguments here

    def __init__(self, *args, **kwargs):
        ...

    def init_states(self):
        ...

    def update_states(self, states, new_state):    
        ...

    def finish_eval(self, states, new_state):
        ...
    
    def policy(self, states):
        ...

    def reset(self):
        ...

```

Here are the implementations of agents for [text *wait-k* model](../eval/agents/simul_trans_text_agent.py) and [speech *wait-k* model](../eval/agents/simul_trans_speech_agent.py).

> Once there is implementation of the agent, to start the evaluation, 

Reviewer: ```suggestion
Once the agent is implemented, to start the evaluation, 
```

```
python $fairseq_dir/examples/simultanesous_translation/eval/evaluate.py \
    --port $port \
    --agent-type my_agent_name \
    --reset-server \
    --num-threads 4 \
    --scores \
    --model-args ... # defined in MyAgent.add_args, such as model path, tokenizer etc.
```

> It can be very slow to evaluate speech model utterence by utterence, please see [here](../scripts/start-multi-client.sh) for a faster implementation, which split the evaluation set into chunks.

Reviewer: ```suggestion
It can be very slow to evaluate the speech model utterance by utterance. See [here](../scripts/start-multi-client.sh) for a faster implementation, which split the evaluation set into chunks.
```

### Quality

The quality is measured by detokenized BLEU. So make sure that the predicted words sent to server are detokenized.

### Latency
The latency metrics are 
* Average Proportion
* Average Lagging
* Differentiable Average Lagging

> For text, the latency will be evaluated on detokenized text. For speech, the will be evaluated based one millisecond

Reviewer: What does `based one millisecond` mean?
